cloud/scope/clients.go

```diff
@@ -17,60 +17,44 @@ limitations under the License.
 package scope
 
 import (
-	"fmt"
 	"strings"
 
 	"github.com/Azure/go-autorest/autorest"
-	"github.com/Azure/go-autorest/autorest/azure"
 	"github.com/Azure/go-autorest/autorest/azure/auth"
 )
 
-const (
-	// ChinaCloud is the cloud environment operated in China
-	ChinaCloud = "AzureChinaCloud"
-	// GermanCloud is the cloud environment operated in Germany
-	GermanCloud = "AzureGermanCloud"
-	// PublicCloud is the default public Azure cloud environment
-	PublicCloud = "AzurePublicCloud"
-	// USGovernmentCloud is the cloud environment for the US Government
-	USGovernmentCloud = "AzureUSGovernmentCloud"
-)
-
 // AzureClients contains all the Azure clients used by the scopes.
 type AzureClients struct {
-	Authorizer                 autorest.Authorizer
-	environment                string
+	Authorizer autorest.Authorizer
+	auth.EnvironmentSettings
 	ResourceManagerEndpoint    string
 	ResourceManagerVMDNSSuffix string
-	subscriptionID             string
-	tenantID                   string
-	clientID                   string
-	clientSecret               string
 }
 
 // CloudEnvironment returns the Azure environment the controller runs in.
 func (c *AzureClients) CloudEnvironment() string {
-	return c.environment
-}
-
-// SubscriptionID returns the Azure subscription id from the controller environment
-func (c *AzureClients) SubscriptionID() string {
-	return c.subscriptionID
+	return c.Environment.Name
 }
 
 // TenantID returns the Azure tenant id the controller runs in.
 func (c *AzureClients) TenantID() string {
-	return c.tenantID
+	return c.Values[auth.TenantID]
 }
 
 // ClientID returns the Azure client id from the controller environment
 func (c *AzureClients) ClientID() string {
-	return c.clientID
+	return c.Values[auth.ClientID]
 }
 
 // ClientSecret returns the Azure client secret from the controller environment
 func (c *AzureClients) ClientSecret() string {
-	return c.clientSecret
+	return c.Values[auth.ClientSecret]
+}
+
+// SubscriptionID returns the Azure subscription id of the cluster,
+// either specified or from the environment
+func (c *AzureClients) SubscriptionID() string {
```

Comment:
Can/should we also remove the SubscriptionID() getter in cluster.go https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/879/files#diff-7c0af1847b5ee759aadb398ac16c4d36L92-L93 ?

Reply:
I think we specifically need that one so we can use the SubscriptionAuthorizer interface, but it should just pass through to the AzureClients (we can't let it use the embedded interface because we already have two methods embedded with that name, from AzureClients and ClusterDescriber, so it's ambigous without explicitly overriding).

Reply:
Well, we don't need it here I guess 🙂 since cluster describer doesn't exist yet